src/Kestrel.Core/Internal/Http2/Http2Connection.cs

```diff
@@ -639,14 +633,15 @@ private Task ProcessSettingsFrameAsync()
                 var previousInitialWindowSize = (int)_clientSettings.InitialWindowSize;
                 var previousMaxFrameSize = _clientSettings.MaxFrameSize;
 
-                _clientSettings.Update(_incomingFrame.GetSettings());
+                _clientSettings.Update(Http2FrameReader.ReadSettings(payload));
 
                 // Ack before we update the windows, they could send data immediately.
-                var ackTask =  _frameWriter.WriteSettingsAckAsync();
+                var ackTask = _frameWriter.WriteSettingsAckAsync();
 
                 if (_clientSettings.MaxFrameSize != previousMaxFrameSize)
                 {
-                    _frameWriter.UpdateMaxFrameSize(_clientSettings.MaxFrameSize);
+                    // Don't let the client choose an arbitrarily large size, this will be used for response buffers.
+                    _frameWriter.UpdateMaxFrameSize(Math.Min(_clientSettings.MaxFrameSize, _serverSettings.MaxFrameSize));
```

Comment:
Can we add a new test to ensure the minimum of the two values is always picked?

Reply:
Added